src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -894,118 +925,344 @@ private bool CheckPropertyValueKind(SyntaxNode node, BoundExpression expr, BindV
                 {
-                    var indexer = indexerAccess.Indexer;
+                    RefKind effectiveRefKind = GetEffectiveRefKind(argIndex, argRefKinds, parameters, argsToParamsOpt, ref inParametersMatchedWithArgs);
 
-                    return CheckArgumentsReturnable(
-                        propertySyntax,
-                        indexer,
-                        indexer.Parameters,
-                        indexerAccess.Arguments,
-                        indexerAccess.ArgumentRefKindsOpt,
-                        indexerAccess.ArgsToParamsOpt,
-                        checkingReceiver,
-                        diagnostics);
+                    // if ref is passed and it is not a ref-like, check ref escape, 
+                    // since it is the same or narrower than val escape and callee may wrap the ref into a val
+                    var argument = args[argIndex];
+                    var argEscape = effectiveRefKind != RefKind.None && argument.Type.IsByRefLikeType == false ?
+                                        GetRefEscape(args[argIndex], scopeOfTheContainingExpression) :
+                                        GetValEscape(args[argIndex], scopeOfTheContainingExpression);
+
+                    escapeScope = Math.Max(escapeScope, argEscape);
+
+                    if (escapeScope >= scopeOfTheContainingExpression)
+                    {
+                        // no longer needed
+                        inParametersMatchedWithArgs?.Free();
+
+                        // can't get any worse
+                        return escapeScope;
+                    }
                 }
             }
 
-            return true;
+            // handle omitted optional "in" parameters if there are any
+            ParameterSymbol unmatchedInParameter = TryGetUnmatchedInParameterAndFreeMatchedArgs(parameters, ref inParametersMatchedWithArgs);
+
+            // unmatched "in" parameter is same as a literal, its ref escape is scopeOfTheContainingExpression  (can't get any worse)
+            //                                                its val escape is ExternalScope                   (does not affect overal result)
+            if (unmatchedInParameter != null)
+            {
+                return scopeOfTheContainingExpression;
+            }
+
+
+            // check receiver if ref-like
+            if (receiverOpt?.Type?.IsByRefLikeType == true)
+            {
+                return GetValEscape(receiverOpt, scopeOfTheContainingExpression);
+            }
+
+            return escapeScope;
         }
 
-        private bool CheckArgumentsReturnable(
+        /// <summary>
+        /// Validates whether given invocation can allow its results to escape from `escapeFrom` level to `escapeTo` level.
+        /// The result indicates whether the escape is possible. 
+        /// Additionally, the method emits diagnostics (possibly more than one, recusrively) that would help identify the cause for the failure.
```

Comment:
Typo (several instances): `recursively`

Reply:
fixed in all 3 places

---
In reply to: [135183538](https://github.com/dotnet/roslyn/pull/21334#discussion_r135183538) [](ancestors = 135183538)